ts/private/ts_project_worker.js

```diff
@@ -0,0 +1,384 @@
+    const builder = ts.createEmitAndSemanticDiagnosticsBuilderProgram(rootNames, options, host, oldProgram, configFileParsingDiagnostics, projectReferences);
+    /** @type {Map} */
+    const emittedFiles = host.emittedFiles = host.emittedFiles || new Map();
+    /** @type {Map} */
+    const emittingMap = host.emittedFilesWeak = host.emittedFilesWeak || new Map();
+
+    const emit = builder.emit;
+    builder.emit = (targetSourceFile, writeFile, cancellationToken, emitOnlyDtsFiles, customTransformers) => {
+        writeFile = writeFile || host.writeFile
+        if (!targetSourceFile) {
+            for (const [path, content] of emittedFiles.entries()) {
+                const sourcePath = emittingMap.get(path);
+                if(!builder.getSourceFile(sourcePath)) {
+                    debuglog(`removing ${path} from emit cache`);
+                    emittedFiles.delete(path);
+                    emittingMap.delete(path);
+                } else {
+                    writeFile(path, content);
+                }
+            }
+        }
+
+        const writeF = (fileName, data, writeByteOrderMark, onError, sourceFiles) => {
+            writeFile(fileName, data, writeByteOrderMark, onError, sourceFiles);
+            emittingMap.set(fileName, sourceFiles[0].fileName)
+            emittedFiles.set(fileName, data)
+        }
+        return emit(targetSourceFile, writeF, cancellationToken, emitOnlyDtsFiles, customTransformers)
+    }
+
+    return builder;
+}
+
+/** @type {ts.FormatDiagnosticsHost} */
+const formatDiagnosticHost = {
+    getCanonicalFileName: (path) => path,
+    getCurrentDirectory: ts.sys.getCurrentDirectory,
+    getNewLine: () => ts.sys.newLine,
+}
+
+function printDiagnostics(diagnostics) {
+    worker.log('')
+    worker.log(ts.formatDiagnostics(diagnostics, formatDiagnosticHost))
+}
+
+
+
+function isAncestorDirectory(ancestor, child) {
+    const ancestorChunks = ancestor.split(path.sep).filter(i => !!i);
+    const childChunks = child.split(path.sep).filter(i => !!i);
+    return ancestorChunks.every((chunk, i) => childChunks[i] === chunk)
+}
+
+/** @type {Map<string, {program: ts.EmitAndSemanticDiagnosticsBuilderProgram, host: ts.WatchCompilerHost, previousInputs?: import("@bazel/worker").Inputs}>} */
+const workers = new Map(); 
+
+function getTsConfigPath(args) {
+    return args[args.indexOf('--project') + 1]
+}
+
+function createProgram(args, initialInputs) {
+    const tsconfig = getTsConfigPath(args);
+    const execRoot = path.resolve(process.cwd(), '..', '..', '..')
+
+    debuglog(`tsconfig: ${tsconfig}`);
+    debuglog(`execroot ${execRoot}`);
+
+    const parsedArgs = ts.parseCommandLine(args)
+
+    const directoryWatchers = new Map();
+    const fileWatchers = new Map();
+
+    const knownInputs = new Set(initialInputs);
+
+    /** @type {ts.System} */
+    const sys = {
+        ...ts.sys,
```

Comment:
FWIW for anyone curious (such as myself)... and @thesayyn let me know if any of this is wrong or you learnt anything else interesting...

https://github.com/microsoft/TypeScript/blob/v4.7.3/src/compiler/sys.ts#L1253 is where this global `sys` object is defined. It seems like there's one and only one instance of this object and that single instance is passed around to almost every method? The global only gets accessed in a couple odd places so passing it into a method essentially overrides it for that unique invocation like what we're doing here.

Here we are cloning it per program to override some methods with caching based on that project.

It also seems to have no prototype or use of `this.` so it's ok to simply clone it instead of something such as `Object.create` to inherit from it?

Reply:
Your findings are accurate. I'd somewhat not complicate things by having only one instance passed around with different binds. I am not sure how performant that would be.  

However, my opinion is not the only one that counts.

Reply:
https://www.npmjs.com/package/@typescript/vfs  is one example that creates more than one sys instances.